lsp/src/uri_lexer.mll

```diff
@@ -1,38 +1,50 @@
 {
 type t =
-  { scheme : string option
+  { scheme : string
   ; authority : string
   ; path : string
   }
 }
 
-rule path = parse
-| '/'? { path1 (Buffer.create 12) lexbuf }
-and path1 buf = parse
-| '\\' { Buffer.add_char buf '/' ; path1 buf lexbuf }
-| "%5" ['c' 'C'] { Buffer.add_char buf '/' ; path1 buf lexbuf }
-| "%3" ['a' 'A'] { Buffer.add_char buf ':' ; path1 buf lexbuf }
-| "%3" ['d' 'D'] { Buffer.add_char buf '=' ; path1 buf lexbuf }
-| "%3" ['f' 'F'] { Buffer.add_char buf '?' ; path1 buf lexbuf }
-| "%20" { Buffer.add_char buf ' ' ; path1 buf lexbuf }
-| _ as c { Buffer.add_char buf c ; path1 buf lexbuf }
-| eof { Buffer.contents buf }
+rule uri = parse
+([^':''/''?''#']+ as scheme ':') ?
+("//" ([^'/''?''#']* as authority)) ?
+([^'?''#']* as path)
+{ 
+  let open Import in
+  let scheme = scheme |> Option.value ~default:"file" in
+  let authority =
+    authority |> Option.map Uri.pct_decode |> Option.value ~default:""
+  in
+  let path =
+    let path = path |> Uri.pct_decode in
+    match scheme with
+    | "http" | "https" | "file" ->
+      String.add_prefix_if_not_exists path ~prefix:"/"
+    | _ -> path
+  in
+  { scheme; authority; path; } 
+}
 
-and uri = parse
-| ([^ ':']+) as scheme ':' { uri1 (Some scheme) lexbuf }
-| "" { uri1 None lexbuf }
-and uri1 scheme = parse
-| "//" ([^ '/']* as authority) { uri2 scheme authority lexbuf }
-| "" { uri2 scheme "" lexbuf }
-and uri2 scheme authority = parse
-| "" { { scheme ; authority ; path = path lexbuf } }
+and path = parse
+| "" { { scheme = "file"; authority = ""; path = "/" } }
+| "//" ([^ '/']* as authority) (['/']_* as path) { { scheme = "file"; authority; path } }
+| "//" ([^ '/']* as authority) { { scheme = "file"; authority; path = "/" } }
+| _* as path 
```

Comment:
What do we need this case for?

Reply:
To cover the case when the string is not empty and doesn't start with `//`. That's the equivalent to https://github.com/ocaml/ocaml-lsp/pull/739/commits/786178025795fd09f95cde61a93e27fc185c05b4#diff-4e2de9533ce99abc9d651d3c346c050c2e1923251c6ad49528deb91519d8748fL48

Also, if I remove that line I get the following error:

```
File "lsp/src/uri_lexer.mll", line 49, character 0: syntax error.
```

Do I misunderstand something?

Reply:
no, I think you got it. Can you split it into two cases though? One with the `/` prefix and one without the `/`. Should remove the needed for any helper functions.

Reply:
Clever. Amazing how this `of_path` function is now way simpler! 🤩